### PR TITLE
Thumbs up/down sparkle component in ConversationMessagesActions

### PR DESCRIPTION
**Pull request by albandum**
## Description

Add a Thumbs up/down sparkle component in ConversationMessagesActions

https://github.com/user-attachments/assets/ea5fc621-8243-40b1-953b-9663090b38b5

### CODE REVIEW

**Comment by spolu** [comment]
Not a designer but would be super nice to have positive feedback also IMHO (good data for future us).

Obviously we don't want a full side modal when we click on :+1: but arguably you can make the argument that we don't want one either when clicking on :-1: since it's not too different in the end?

Do we have ideas to have something more localized and less disruptive but that will bait you into doing it?

File: sparkle/src/components/ConversationMessageActions.tsx [hunk 1]
```diff
@@ -174,3 +203,98 @@ function EmojiSelector({
...
+
+  return (
+    <div
+      ref={containerRef}
+      className="s-inline-flex s-h-7 s-items-center s-justify-center s-gap-2 s-whitespace-nowrap s-rounded-lg s-border s-border-border-dark s-bg-background s-px-2.5 s-text-xs s-font-medium s-text-primary-dark s-ring-offset-background s-transition-colors hover:s-border-primary-150 hover:s-bg-primary-150 hover:s-text-primary focus-visible:s-outline-none focus-visible:s-ring-2 focus-visible:s-ring-ring focus-visible:s-ring-offset-2 active:s-bg-primary-300 disabled:s-pointer-events-none disabled:s-border-structure-100 disabled:s-text-primary-muted"
+    >
+      <button
+        disabled={isSubmittingThumb}
+        onClick={() => selectThumb("up")}
+        className={`p-1.5 disabled:opacity-50 disabled:cursor-not-allowed hover:s-text-blue-600 ${selectedThumb === "up" ? "s-text-blue-600" : ""}`}
```
Comment (JulesBelveze) [inline sparkle/src/components/ConversationMessageActions.tsx:256]:
```suggestion
        className={`s-p-1.5 disabled:s-opacity-50 disabled:s-cursor-not-allowed hover:s-text-blue-600 ${selectedThumb === "up" ? "s-text-blue-600" : ""}`}
```

File: sparkle/src/components/ConversationMessageActions.tsx [hunk 2]
```diff
@@ -174,3 +203,98 @@ function EmojiSelector({
...
+      >
+        <HandThumbUpIcon className="h-4 w-4" />
```
Comment (JulesBelveze) [inline sparkle/src/components/ConversationMessageActions.tsx:258]:
```suggestion
        <HandThumbUpIcon className="s-h-4 s-w-4" />
```

File: sparkle/src/components/ConversationMessageActions.tsx [hunk 3]
```diff
@@ -174,3 +203,98 @@ function EmojiSelector({
...
+        isOpen={isThumbDownModalOpened}
+        onClose={() => setIsThumbDownModalOpened(false)}
+        hasChanged={false}
+        variant="side-sm"
+        title="Feedback"
+      >
+        <div className="py-8">
+          <TextArea
+            placeholder="What was unsatisfactory about this answer?"
+            className="mt-4"
```
Comment (JulesBelveze) [inline sparkle/src/components/ConversationMessageActions.tsx:277]:
```suggestion
            className="s-mt-4"
```

File: sparkle/src/components/ConversationMessageActions.tsx [hunk 4]
```diff
@@ -174,3 +203,98 @@ function EmojiSelector({
...
+        <HandThumbDownIcon className="h-4 w-4" />
+      </button>
+      <Modal
+        isOpen={isThumbDownModalOpened}
+        onClose={() => setIsThumbDownModalOpened(false)}
+        hasChanged={false}
+        variant="side-sm"
+        title="Feedback"
+      >
+        <div className="py-8">
```
Comment (JulesBelveze) [inline sparkle/src/components/ConversationMessageActions.tsx:274]:
```suggestion
        <div className="s-py-8">
```

File: sparkle/src/components/ConversationMessageActions.tsx [hunk 5]
```diff
@@ -174,3 +203,98 @@ function EmojiSelector({
...
+        className={`p-1.5 disabled:opacity-50 disabled:cursor-not-allowed hover:s-text-blue-600 ${selectedThumb === "up" ? "s-text-blue-600" : ""}`}
+      >
+        <HandThumbUpIcon className="h-4 w-4" />
+      </button>
+      <button
+        disabled={isSubmittingThumb}
+        onClick={() => selectThumb("down")}
+        className={`p-1.5 disabled:opacity-50 disabled:cursor-not-allowed hover:s-text-blue-600 ${selectedThumb === "down" ? "s-text-blue-600" : ""}`}
+      >
+        <HandThumbDownIcon className="h-4 w-4" />
```
Comment (JulesBelveze) [inline sparkle/src/components/ConversationMessageActions.tsx:265]:
```suggestion
        <HandThumbDownIcon className="s-h-4 s-w-4" />
```

File: sparkle/src/components/ConversationMessageActions.tsx [hunk 6]
```diff
@@ -174,3 +203,98 @@ function EmojiSelector({
...
+        disabled={isSubmittingThumb}
+        onClick={() => selectThumb("up")}
+        className={`p-1.5 disabled:opacity-50 disabled:cursor-not-allowed hover:s-text-blue-600 ${selectedThumb === "up" ? "s-text-blue-600" : ""}`}
+      >
+        <HandThumbUpIcon className="h-4 w-4" />
+      </button>
+      <button
+        disabled={isSubmittingThumb}
+        onClick={() => selectThumb("down")}
+        className={`p-1.5 disabled:opacity-50 disabled:cursor-not-allowed hover:s-text-blue-600 ${selectedThumb === "down" ? "s-text-blue-600" : ""}`}
```
Comment (JulesBelveze) [inline sparkle/src/components/ConversationMessageActions.tsx:263]:
```suggestion
        className={`s-p-1.5 disabled:s-opacity-50 disabled:s-cursor-not-allowed hover:s-text-blue-600 ${selectedThumb === "down" ? "s-text-blue-600" : ""}`}
```

**Review by JulesBelveze — COMMENTED** [review]
I might have forgotten a bunch of prefixes, mind fixing them? I'll have a second look at it afterwards 🤓

**Comment by albandum** [comment]
Decision (https://github.com/dust-tt/decisions/issues/296) was taken to only show the text input on bad feedback. Letting @pinotalexandre answer that

Also second question to ask is: as we want to force text feedback on bad feedback, do we still register bad feedbacks _before_ the modal (on thumb click) and then update the feedback with the message, right ? (meaning we don't _actually_ force anything?

There's some UX thought to be given around that, because if there's a misclick, the modal opens, I close the modal - I might not see that the 👎 became blue and that my feedback was recorded 🤔

**Comment by spolu** [comment]
> Also second question to ask is: as we want to force text feedback on bad feedback, do we still register bad feedbacks before the modal (on thumb click) and then update the feedback with the message, right ? (meaning we don't actually force anything?

Definitely

> There's some UX thought to be given around that, because if there's a misclick, the modal opens, I close the modal - I might not see that the 👎 became blue and that my feedback was recorded 🤔

We simply want the feedback to be sticky so that I can change it after no?

**Comment by pinotalexandre** [comment]
@albandum 
We misunderstood each other, I think only negative feedback should be mandatory, but still allow the option to leave written positive feedback :) 

I edited the feedback loop issue : 
image

**Comment by albandum** [comment]
Icon size update 
Screenshot 2024-11-21 at 14 53 03

**Comment by pinotalexandre** [comment]
Much better!

**Comment by pinotalexandre** [comment]
I think @spolu is right about the side modal. 
Here the modal "interrupt" the users flow. Imo, the effort we are putting into creating a floating dialog would be rewarded by a higher number of feedback.

About the mandatory written feedback for thumbs down. Let's nudge the user by saying that leaving a written feedback is mandatory, but If he gives up, let's take into account the thumbs down.

@albandum

**Comment by albandum** [comment]
https://github.com/user-attachments/assets/7b6a1ab4-4fa4-4948-994b-f369c8c4d152

@pinotalexandre ?

File: sparkle/src/components/ConversationMessageActions.tsx [hunk 7]
```diff
@@ -174,3 +208,111 @@ function EmojiSelector({
     />
   );
 }
+
+function ConversationMessageThumbsSelector({
+  onSubmitThumb,
+  isSubmittingThumb,
+}: ConversationMessageThumbSelectorProps) {
+  return (
+    <>
```
Comment (JulesBelveze) [inline sparkle/src/components/ConversationMessageActions.tsx:216]: `nit`: do we need the fragment here?

File: sparkle/src/components/ConversationMessageActions.tsx [hunk 8]
```diff
@@ -174,3 +208,111 @@ function EmojiSelector({
...
+      setIsPopoverOpen(false);
+      await onSubmitThumb({ thumb, isToRemove: true });
+      return;
+    }
+    setSelectedThumb(thumb);
+    setIsPopoverOpen(true);
+    await onSubmitThumb({ thumb, isToRemove: false });
+  };
+
+  const [selectedThumb, setSelectedThumb] =
```
Comment (JulesBelveze) [inline sparkle/src/components/ConversationMessageActions.tsx:243]: `nit`: let's move all the hooks at the top of the function

File: sparkle/src/components/ConversationMessageActions.tsx [hunk 9]
```diff
@@ -174,3 +207,110 @@ function EmojiSelector({
...
+        <PopoverTrigger>
+          <div className="s-flex s-items-center">
+            <Button
+              variant="outline"
+              size="xs"
+              disabled={isSubmittingThumb}
+              onClick={() => selectThumb("up")}
+              className={`s-rounded-r-none s-border-r-0 ${
+                selectedThumb === "up" ? "s-text-blue-600" : ""
+              }`}
```
Comment (JulesBelveze) [inline sparkle/src/components/ConversationMessageActions.tsx:260]:
```suggestion
              className={cn("s-rounded-r-none s-border-r-0", 
                selectedThumb === "up" ? "s-text-blue-600" : ""
              )}
```

Comment (JulesBelveze) [reply sparkle/src/components/ConversationMessageActions.tsx:260]: We usually go for this pattern, could you the same below?

File: sparkle/src/components/ConversationMessageActions.tsx [hunk 10]
```diff
@@ -174,3 +207,110 @@ function EmojiSelector({
...
+
+  const [selectedThumb, setSelectedThumb] =
+    React.useState<ThumbReaction | null>(null);
+  const [feedback, setFeedback] = React.useState<string | null>(null);
+  const [isPopoverOpen, setIsPopoverOpen] = React.useState(false);
+
+  return (
+    <div ref={containerRef} className="s-flex s-items-center">
+      <PopoverRoot open={isPopoverOpen}>
+        <PopoverTrigger>
```
Comment (JulesBelveze) [inline sparkle/src/components/ConversationMessageActions.tsx:251]:
```suggestion
        <PopoverTrigger asChild>
```

**Review by JulesBelveze — COMMENTED** [review]
Looking good, left a few coms. Feel free to merge once addressed 👌🏼

File: sparkle/src/components/ConversationMessageActions.tsx [hunk 11]
```diff
@@ -174,3 +207,110 @@ function EmojiSelector({
...
+              icon={HandThumbUpIcon}
+            />
+            <Button
+              variant="outline"
+              size="xs"
+              disabled={isSubmittingThumb}
+              onClick={() => selectThumb("down")}
+              className={`s-rounded-l-none s-border-l-0 ${
+                selectedThumb === "down" ? "s-text-blue-600" : ""
+              }`}
```
Comment (JulesBelveze) [inline sparkle/src/components/ConversationMessageActions.tsx:270]: Could you do the same here?

**Comment by albandum** [comment]
https://github.com/user-attachments/assets/f0f81b3e-76ca-4fcc-8044-db0d3d053a41

**Comment by albandum** [comment]
https://github.com/user-attachments/assets/b5ec2446-1674-4290-98fc-b1b3b570e044

**Pull request merged**